File: docs-examples/example-java/src/test/java/io/micronaut/nats/docs/consumer/connection/ConnectionSpec.java
package io.micronaut.nats.docs.consumer.connection;

import io.micronaut.context.annotation.Property;
import io.micronaut.nats.testcontainers.Nats;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@MicronautTest
@Property(name = "spec.name", value = "ConnectionSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ConnectionSpec implements TestPropertyProvider {

    @Test
    void testProductClientAndListener(ProductClient productClient, ProductListener productListener) {

        // tag::producer[]
        productClient.send("connection-test".getBytes());
        // end::producer[]

        await().atMost(10, SECONDS).until(() ->
                productListener.messageLengths.size() == 1 &&
                productListener.messageLengths.get(0).equals("connection-test")
        );
    }

    @Override
    public Map<String, String> getProperties() {
        Map<String, String> containerProps = Nats.getProperties();
        String natsAddress = containerProps.get("nats.addresses");
        return Map.of(
            "nats.product-cluster.addresses", natsAddress
        );
    }
}

File: docs-examples/example-python/src/test/python/micronaut/nats/docs/consumer/connection/ConnectionSpec.py
from time import sleep
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@Property(name="spec.name", value="ConnectionSpec")
@Property(name="nats.product-cluster.addresses", value="${nats.addresses}")
@MicronautTest(environments=["nats"])
class ConnectionSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_product_client_and_listener(self):
        # tag::producer[]
        self.product_client.send("connection-test".encode())
        # end::producer[]

        for _ in range(100):
            if self.product_listener.message_lengths == ["connection-test"]:
                break
            sleep(0.1)
        assert self.product_listener.message_lengths == ["connection-test"]
